HW_2.2/hw_2.py

```diff
@@ -0,0 +1,342 @@
+
+**Модули использовать нельзя**
+"""
+
+# Ваш код здесь
+class MyDict(dict):
+    def __iter__(self):
+        for key in super().__iter__():
+            yield key, self[key]
+
+dct = MyDict({"a": 1, "b": 2, "c": 3, "d": 25})
+for key, value in dct:
+    print(key, value)
+
+for key, value in dct.items():
+    print(key, value)
+
+for key in dct.keys():
+    print(key)
+
+dct["c"] + dct["d"]
+
+"""# Задание 2 (2 балла)
+
+Напишите функцию `iter_append`, которая "добавляет" новый элемент в конец итератора, возвращая итератор, который включает изначальные элементы и новый элемент. Итерироваться по итератору внутри функции нельзя, то есть вот такая штука не принимается
+```python
+def iter_append(iterator, item):
+    lst = list(iterator) + [item]
+    return iter(lst)
+```
+
+**Модули использовать нельзя**
+"""
+
+def iter_append(iterator, item):
+    for i in iterator:
+        yield i
+    yield item
+    
+
+my_iterator = iter([1, 2, 3])
+new_iterator = iter_append(my_iterator, 4)
+
+for element in new_iterator:
+    print(element)
+    break
+
+"""# Задание 3 (5 баллов)
+
+Представим, что мы установили себе некотурую библиотеку, которая содержит в себе два класса `MyString` и `MySet`, которые являются наследниками `str` и `set`, но также несут и дополнительные методы.
+
+Проблема заключается в том, что библиотеку писали не очень аккуратные люди, поэтому получилось так, что некоторые методы возвращают не тот тип данных, который мы ожидаем. Например, `MyString().reverse()` возвращает объект класса `str`, хотя логичнее было бы ожидать объект класса `MyString`.
+
+Найдите и реализуйте удобный способ сделать так, чтобы подобные методы возвращали экземпляр текущего класса, а не родительского. При этом **код методов изменять нельзя**
+
+**+3 дополнительных балла** за реализацию того, чтобы **унаследованные от `str` и `set` методы** также возвращали объект интересующего нас класса (то есть `MyString.replace(..., ...)` должен возвращать `MyString`). **Переопределять методы нельзя**
+
+**Модули использовать нельзя**
+"""
+
+# Ваш код где угодно, но не внутри методов
+def wrap_class(cls):
+    
+    def decor_function( attr):
+        def wrap_result(*args, **kwargs):
+            res = attr(*args, **kwargs)
+            if isinstance(res, str):
+                return MyString(res)
+            if isinstance(res, set):
+                return MySet(res)
+            return res
+        return wrap_result
+    
+    class CustomClassWrapper(cls):
+        def __init__(self, *args, **kwargs):
+            for attr_name in dir(self):
+                if not attr_name.startswith('__') and not attr_name.endswith('__'):
+                    attr = getattr(self, attr_name)
+                    if callable(attr):
+                        setattr(self, attr_name, decor_function(attr))
+    return CustomClassWrapper
```

Comment:
```suggestion
def decor_function( attr):
    def wrap_result(*args, **kwargs):
        res = attr(*args, **kwargs)
        if isinstance(res, str):
            return MyString(res)
        if isinstance(res, set):
            return MySet(res)
        return res
    return wrap_result

def CustomClassWrapper(cls):
    for attr_name in dir(cls):
        if not attr_name.startswith('__') and not attr_name.endswith('__'):
            attr = getattr(cls, attr_name)
            if callable(attr):
                setattr(cls, attr_name, decor_function(attr))
    return cls
```

И CustomClassWrapper на классы.
В текущей реализации примеры из задания возвращают CustomClassWrapper - это вообще не ожидаемое поведение.